Code/GraphMol/RascalMCES/RascalMCES.cpp

```diff
@@ -0,0 +1,1086 @@
+      break;
+    }
+  }
+}
+
+bool deltaYExchangePossible(const ROMol &mol1, const ROMol &mol2) {
+  // A Delta-y exchange is an incorrect match when a cyclopropyl ring (the
+  // delta) is matched to a C(C)(C) group (the y) because they both have
+  // isomorphic line graphs.  This checks to see if that's something we need to
+  // worry about for these molecules.
+  const static std::unique_ptr<ROMol> delta(SmartsToMol("C1CC1"));
+  const static std::unique_ptr<ROMol> y(SmartsToMol("C(C)C"));
+  MatchVectType dontCare;
+  return (SubstructMatch(mol1, *delta, dontCare) &&
+          SubstructMatch(mol2, *y, dontCare)) ||
+         (SubstructMatch(mol2, *delta, dontCare) &&
+          SubstructMatch(mol1, *y, dontCare));
+}
+
+void findEquivalentBonds(const ROMol &mol, std::vector<int> &equivBonds) {
+  equivBonds = std::vector<int>(mol.getNumBonds(), -1);
+  std::vector<unsigned int> ranks(mol.getNumAtoms());
+  bool breakTies = false;
+  Canon::rankMolAtoms(mol, ranks, breakTies);
+  int nextClass = 0;
+  for (const auto &b1 : mol.bonds()) {
+    for (const auto &b2 : mol.bonds()) {
+      if (b1->getIdx() != b2->getIdx()) {
+        if ((ranks[b1->getBeginAtomIdx()] == ranks[b2->getBeginAtomIdx()] &&
+             ranks[b1->getEndAtomIdx()] == ranks[b2->getEndAtomIdx()]) ||
+            (ranks[b1->getBeginAtomIdx()] == ranks[b2->getEndAtomIdx()] &&
+             ranks[b1->getEndAtomIdx()] == ranks[b2->getBeginAtomIdx()])) {
+          if (equivBonds[b1->getIdx()] == -1 &&
+              equivBonds[b2->getIdx()] == -1) {
+            equivBonds[b1->getIdx()] = nextClass;
+            equivBonds[b2->getIdx()] = nextClass;
+            ++nextClass;
+          } else if (equivBonds[b1->getIdx()] == -1) {
+            equivBonds[b1->getIdx()] = equivBonds[b2->getIdx()];
+          } else if (equivBonds[b2->getIdx()] == -1) {
+            equivBonds[b2->getIdx()] = equivBonds[b1->getIdx()];
+          }
+        }
+      }
+    }
+  }
+}
+
+// Use the Floyd-Warshall algorithm to compute the distance matrix from the
+// adjacency matrix.
+// Adapted from https://en.wikipedia.org/wiki/Floyd–Warshall_algorithm
+void calcDistMatrix(const std::vector<std::vector<int>> &adjMatrix,
+                    std::vector<std::vector<int>> &distMatrix) {
+  distMatrix = std::vector<std::vector<int>>(
+      adjMatrix.size(),
+      std::vector<int>(adjMatrix.size(), adjMatrix.size() + 1));
+  for (size_t i = 0u; i < adjMatrix.size(); ++i) {
+    distMatrix[i][i] = 0;
+    for (size_t j = 0u; j < adjMatrix.size(); ++j) {
+      if (i != j && adjMatrix[i][j]) {
+        distMatrix[i][j] = 1;
+      }
+    }
+  }
+  for (size_t k = 0u; k < adjMatrix.size(); ++k) {
+    for (size_t i = 0u; i < adjMatrix.size(); ++i) {
+      for (size_t j = 0u; j < adjMatrix.size(); ++j) {
+        if (distMatrix[i][j] > distMatrix[i][k] + distMatrix[k][j]) {
+          distMatrix[i][j] = distMatrix[i][k] + distMatrix[k][j];
+        }
+      }
+    }
+  }
+}
```

Comment:
There's already a floyd-warshall implementation in `MolOps::getDistanceMat()`
Maybe it makes sense to use that?

Reply:
Don't those work off the molecule, though?  Here I'm doing it for the line graph - the graph where the nodes are bonds and the edges are atoms.